File: src/hilbert/sources/semantic_scholar.py
"""Semantic Scholar API client for Hilbert."""

import os
from datetime import datetime
from typing import Any, Dict, List, Optional
import aiohttp

from hilbert.models.paper import Author, Paper


SEMANTIC_SCHOLAR_API_URL = "https://api.semanticscholar.org/graph/v1"


class SemanticScholarError(Exception):
    """Semantic Scholar API error."""
    pass


class SemanticScholarClient:
    """Semantic Scholar API client."""

    MAX_RESULTS = 25

    def __init__(self, api_key: Optional[str] = None, max_results: int = 25):
        self.api_key = api_key or os.getenv("SEMANTIC_SCHOLAR_API_KEY")
        self.max_results = max_results
        self._session: Optional[aiohttp.ClientSession] = None

    async def _get_session(self) -> aiohttp.ClientSession:
        """Get or create aiohttp session."""
        if self._session is None or self._session.closed:
            headers = {}
            if self.api_key:
                headers["x-api-key"] = self.api_key
            self._session = aiohttp.ClientSession(headers=headers)
        return self._session

    async def close(self) -> None:
        """Close the session."""
        if self._session and not self._session.closed:
            await self._session.close()

    def _parse_paper(self, data: Dict[str, Any]) -> Paper:
        """Parse Semantic Scholar paper data."""
        paper_id = data.get("paperId", f"sem-{data.get('paperId', '')}")

        title = data.get("title", "Untitled")
        abstract = data.get("abstract", "")

        authors = []
        for author in data.get("authors", []):
            name = author.get("name", "")
            if name:
                authors.append(Author(name=name))

        year = data.get("year")
        published_date = None
        if year:
            try:
                published_date = datetime.strptime(str(year), "%Y").date()
            except ValueError:
                pass

        url = data.get("url", "")
        if not url and paper_id:
            url = f"https://www.semanticscholar.org/paper/{paper_id}"

        venue = data.get("venue", "")

        doi = data.get("doi", "")

        citation_count = data.get("citationCount", 0)

        is_open = data.get("isOpenAccess", False)

        return Paper(
            paper_id=paper_id,
            title=title,
            abstract=abstract,
            authors=authors,
            published_date=published_date,
            url=url,
            doi=doi,
            venue=venue,
            citation_count=citation_count,
            is_open_access=is_open,
        )

    async def search(
        self,
        query: str,
        max_results: Optional[int] = None,
        year_from: Optional[int] = None,
    ) -> List[Paper]:
        """Search Semantic Scholar for papers.

        Args:
            query: Search terms.
            max_results: Maximum number of results.
            year_from: If given, restrict results to papers published in this
                year or later.  Passed as the ``year`` range parameter
                (e.g. ``year_from=2023`` → ``year=2023-``).
        """
        max_results = max_results or self.max_results

        params: Dict[str, Any] = {
            "query": query,
            "limit": max_results,
            "fields": "title,abstract,authors,year,url,venue,doi,citationCount,isOpenAccess",
        }

        if year_from is not None:
            params["year"] = f"{year_from}-"

        url = f"{SEMANTIC_SCHOLAR_API_URL}/paper/search"

        try:
            session = await self._get_session()
            async with session.get(url, params=params) as response:
                if response.status != 200:
                    text = await response.text()
                    raise SemanticScholarError(f"Semantic Scholar API error: {response.status}: {text}")

                data = await response.json()

            papers = []
            for item in data.get("data", []):
                paper = self._parse_paper(item)
                papers.append(paper)

            return papers

        except SemanticScholarError:
            raise
        except Exception as e:
            raise SemanticScholarError(f"Failed to search Semantic Scholar: {e}") from e

    async def get_paper(self, paper_id: str) -> Optional[Paper]:
        """Get a specific paper by ID."""
        url = f"{SEMANTIC_SCHOLAR_API_URL}/paper/{paper_id}"

        params = {
            "fields": "title,abstract,authors,year,url,venue,doi,citationCount,isOpenAccess",
        }

        try:
            session = await self._get_session()
            async with session.get(url, params=params) as response:
                if response.status == 404:
                    return None
                if response.status != 200:
                    raise SemanticScholarError(f"Semantic Scholar API error: {response.status}")

                data = await response.json()
                return self._parse_paper(data)

        except SemanticScholarError:
            raise
        except Exception as e:
            raise SemanticScholarError(f"Failed to get paper: {e}") from e

    async def get_citations(self, paper_id: str, limit: int = 10) -> List[Paper]:
        """Get papers citing the given paper."""
        url = f"{SEMANTIC_SCHOLAR_API_URL}/paper/{paper_id}/citations"

        params = {
            "limit": limit,
            "fields": "title,abstract,authors,year,url,venue,doi,citationCount",
        }

        try:
            session = await self._get_session()
            async with session.get(url, params=params) as response:
                if response.status != 200:
                    return []

                data = await response.json()

            papers = []
            for cite in data.get("data", []):
                citing_paper = cite.get("citingPaper", {})
                if citing_paper:
                    paper = self._parse_paper(citing_paper)
                    papers.append(paper)

            return papers

        except Exception:
            return []


    async def get_references(self, paper_id: str, limit: int = 20) -> list[str]:
        """Return Semantic Scholar paper IDs that this paper cites (references).

        Returns a list of paperId strings, not full Paper objects, to keep
        this cheap — we only need IDs for the within-corpus citation graph.
        """
        url = f"{SEMANTIC_SCHOLAR_API_URL}/paper/{paper_id}/references"
        params = {"limit": limit, "fields": "paperId"}

        try:
            session = await self._get_session()
            async with session.get(url, params=params) as response:
                if response.status != 200:
                    return []
                data = await response.json()
            return [
                ref["citedPaper"]["paperId"]
                for ref in data.get("data", [])
                if ref.get("citedPaper", {}).get("paperId")
            ]
        except Exception:
            return []


_client: Optional[SemanticScholarClient] = None


def get_semantic_scholar_client() -> SemanticScholarClient:
    """Get Semantic Scholar client singleton."""
    global _client
    if _client is None:
        _client = SemanticScholarClient()
    return _client


def set_semantic_scholar_client(client: SemanticScholarClient) -> None:
    """Set Semantic Scholar client singleton."""
    global _client
    _client = client

File: cli/src/core/sources/semanticScholar.ts
import { getSettings } from "../config";
import type { Paper } from "../models/paper";

const SEMANTIC_SCHOLAR_API_URL = "https://api.semanticscholar.org/graph/v1";

export class SemanticScholarClient {
  private apiKey: string | undefined;
  private maxResults: number;

  constructor(apiKey?: string, maxResults: number = 25) {
    const settings = getSettings();
    this.apiKey = apiKey || settings.semanticScholarApiKey;
    this.maxResults = maxResults;
  }

  private getHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };
    if (this.apiKey) {
      headers["x-api-key"] = this.apiKey;
    }
    return headers;
  }

  private parsePaper(data: Record<string, unknown>): Paper {
    const paperId = (data.paperId as string) || `sem-${data.paperId || ""}`;
    const title = (data.title as string) || "Untitled";
    const abstract = (data.abstract as string) || "";

    const authors: { name: string; affiliation?: string; author_id?: string }[] = [];
    const authorData = data.authors;
    if (Array.isArray(authorData)) {
      for (const a of authorData) {
        if (a && typeof a === "object") {
          const name = (a as Record<string, unknown>).name as string;
          if (name) authors.push({ name });
        }
      }
    }

    let publishedDate: string | undefined;
    const year = data.year;
    if (year) {
      publishedDate = String(year);
    }

    let url = (data.url as string) || "";
    if (!url && paperId) {
      url = `https://www.semanticscholar.org/paper/${paperId}`;
    }

    const venue = (data.venue as string) || "";
    const doi = (data.doi as string) || "";
    const citationCount = (data.citationCount as number) || 0;
    const isOpenAccess = (data.isOpenAccess as boolean) || false;

    return {
      paper_id: paperId,
      title,
      abstract,
      authors,
      published_date: publishedDate,
      url,
      doi: doi || undefined,
      venue: venue || undefined,
      citation_count: citationCount,
      is_open_access: isOpenAccess,
    };
  }

  async search(query: string, maxResults?: number, yearFrom?: number): Promise<Paper[]> {
    const max = maxResults || this.maxResults;

    const params = new URLSearchParams({
      query,
      limit: String(max),
      fields: "title,abstract,authors,year,url,venue,doi,citationCount,isOpenAccess",
    });

    if (yearFrom !== undefined) {
      params.set("year", `${yearFrom}-`);
    }

    const url = `${SEMANTIC_SCHOLAR_API_URL}/paper/search?${params.toString()}`;

    try {
      const response = await fetch(url, { headers: this.getHeaders() });
      if (!response.ok) {
        const text = await response.text();
        throw new Error(`Semantic Scholar API error: ${response.status}: ${text}`);
      }

      const data = (await response.json()) as { data?: Record<string, unknown>[] };
      const items = data?.data;
      if (!items) return [];

      return items.map((item) => this.parsePaper(item));
    } catch (error) {
      throw new Error(`Failed to search Semantic Scholar: ${error}`);
    }
  }

  async getPaper(paperId: string): Promise<Paper | null> {
    const params = new URLSearchParams({
      fields: "title,abstract,authors,year,url,venue,doi,citationCount,isOpenAccess",
    });

    const url = `${SEMANTIC_SCHOLAR_API_URL}/paper/${paperId}?${params.toString()}`;

    try {
      const response = await fetch(url, { headers: this.getHeaders() });
      if (response.status === 404) return null;
      if (!response.ok) {
        throw new Error(`Semantic Scholar API error: ${response.status}`);
      }

      const data = (await response.json()) as Record<string, unknown>;
      return this.parsePaper(data);
    } catch (error) {
      throw new Error(`Failed to get paper: ${error}`);
    }
  }

  async getCitations(paperId: string, limit: number = 10): Promise<Paper[]> {
    const params = new URLSearchParams({
      limit: String(limit),
      fields: "title,abstract,authors,year,url,venue,doi,citationCount",
    });

    const url = `${SEMANTIC_SCHOLAR_API_URL}/paper/${paperId}/citations?${params.toString()}`;

    try {
      const response = await fetch(url, { headers: this.getHeaders() });
      if (!response.ok) return [];

      const data = (await response.json()) as { data?: Record<string, unknown>[] };
      const citations = data?.data;
      if (!citations) return [];

      return citations
        .map((c) => (c as Record<string, unknown>).citingPaper)
        .filter((p): p is Record<string, unknown> => !!p)
        .map((p) => this.parsePaper(p));
    } catch {
      return [];
    }
  }

  async getReferences(paperId: string, limit: number = 20): Promise<string[]> {
    const params = new URLSearchParams({
      limit: String(limit),
      fields: "paperId",
    });

    const url = `${SEMANTIC_SCHOLAR_API_URL}/paper/${paperId}/references?${params.toString()}`;

    try {
      const response = await fetch(url, { headers: this.getHeaders() });
      if (!response.ok) return [];

      const data = (await response.json()) as { data?: Record<string, unknown>[] };
      const refs = data?.data;
      if (!refs) return [];

      return refs
        .map((r) => (r as Record<string, unknown>).citedPaper)
        .filter((p): p is Record<string, unknown> => !!p && typeof p === "object")
        .map((p) => p.paperId as string);
    } catch {
      return [];
    }
  }
}

let clientInstance: SemanticScholarClient | null = null;

export function getSemanticScholarClient(): SemanticScholarClient {
  if (!clientInstance) {
    clientInstance = new SemanticScholarClient();
  }
  return clientInstance;
}

export function setSemanticScholarClient(client: SemanticScholarClient): void {
  clientInstance = client;
}